tc/core/polyhedral/schedule_tree.cc

```diff
@@ -286,7 +286,7 @@ std::unique_ptr<ScheduleTree> ScheduleTree::makeMappingUnsafe(
       << "some id is used more than once in the mapping";
   auto ctx = mappedIds[0].get_ctx();
   auto res =
-      std::unique_ptr<ScheduleTree>(new ScheduleTreeElemMapping(mapping));
+      std::unique_ptr<ScheduleTree>(new ScheduleTreeElemMapping(ctx, mapping));
```

Comment:
do we need 4 different commits for the same change applied to different types, I'd squash

Reply:
Not sure.  The changes are conceptually similar (except for `ThreadSpecificMapping`, which has a different motivation), but completely independent from each other.  What do you think @skimo-openhub ?

Reply:
On Tue, Jul 03, 2018 at 11:58:25AM -0700, ftynse wrote:
> Not sure.  The changes are conceptually similar (except for `ThreadSpecificMapping`, which
has a different motivation), but completely independent from each other.  What do you think @
skimo-openhub ?

I think the current split-up is fine.

However, if you really want to squash things, then I'd squash
all four of them into
"ScheduleTreeElem*: inherit from ScheduleTree instead of ScheduleTreeElemBase"
because I don't think it's good idea to "break" things (at least prompting
you to introduce FIXMEs) and then fixing them up later in the same series.

On the other hand, I'm not sure why we are storing the isl::ctx in schedule
trees in the first place.  Other parts of the code are apparently simply
using the global ctx.  I don't know what the motivation was for introducing
this global ctx (it was introduced by a commit saying only
"Trying to rebase #300 properly after github craziness"), so I can't tell 
what makes schedule trees so special that they shouldn't use the global ctx.

skimo

Reply:
> However, if you really want to squash things, then I'd squash
all four of them into
"ScheduleTreeElem*: inherit from ScheduleTree instead of ScheduleTreeElemBase"
because I don't think it's good idea to "break" things (at least prompting
you to introduce FIXMEs) and then fixing them up later in the same series.

I wouldn't say things are broken, they are just inconsistent.

> On the other hand, I'm not sure why we are storing the isl::ctx in schedule
trees in the first place. Other parts of the code are apparently simply
using the global ctx.

There are very few parts that use the global ctx: compiler API, mapping id constructors and some tests.  Pretty much everywhere else we extract the ctx from an available isl object, or pass it from somewhere.  At this point, I'd say we keep doing that for the sake of consistency.

> I don't know what the motivation was for introducing
this global ctx

AFAIR, in the old "executor" API based on C2 operators, there was a problem with copying objects between threads.  Anyway, a `thread_local` ctx is a reasonable thing to have until there is support for moving objects between threads/ctxs.